File: reference/test/reference_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/mockito.dart';
import 'package:reference/reference.dart';

import 'reference_matchers.dart';

void main() {
  group('$RemoteReferenceMap', () {
    TestReferencePairManager testManager;

    setUp(() {
      testManager = TestReferencePairManager()..initialize();
    });

    test('pairWithNewLocalReference', () {
      when(testManager.localHandler.createInstance(testManager, TestClass, any))
          .thenReturn(TestClass());

      final TestClass result = testManager.onReceiveCreateNewPair(
        RemoteReference('apple'),
        0,
      );

      expect(
        testManager.getPairedLocalReference(RemoteReference('apple')),
        result,
      );
      expect(
        testManager.getPairedRemoteReference(result),
        RemoteReference('apple'),
      );

      verifyInOrder([
        testManager.converter.mock.convertForLocalManager(
          testManager,
          argThat(isEmpty),
        ),
        testManager.localHandler.createInstance(
          testManager,
          TestClass,
          argThat(isEmpty),
        ),
      ]);
    });

    test('pairWithNewLocalReference returns null', () {
      when(testManager.localHandler.createInstance(testManager, TestClass, any))
          .thenReturn(TestClass());

      testManager.onReceiveCreateNewPair(RemoteReference('apple'), 0);
      expect(
        testManager.onReceiveCreateNewPair(RemoteReference('apple'), 0),
        isNull,
      );
    });

    test('invokeLocalStaticMethod', () {
      testManager.onReceiveInvokeStaticMethod(TestClass, 'aStaticMethod');
      verifyInOrder([
        testManager.converter.mock.convertForLocalManager(
          testManager,
          argThat(isNull),
        ),
        testManager.localHandler.sendInvokeStaticMethod(
          testManager,
          TestClass,
          'aStaticMethod',
          argThat(isEmpty),
        ),
        testManager.converter.mock.convertForRemoteManager(
          testManager,
          argThat(isNull),
        ),
      ]);
    });

    test('invokeLocalMethod', () {
      when(testManager.localHandler.createInstance(testManager, TestClass, any))
          .thenAnswer((_) => TestClass());

      final TestClass testClass = testManager.onReceiveCreateNewPair(
        RemoteReference('chi'),
        0,
      );

      testManager.onReceiveInvokeMethod(testClass, 'aMethod');

      verifyInOrder([
        testManager.converter.mock.convertForLocalManager(
          testManager,
          argThat(isNull),
        ),
        testManager.localHandler.sendInvokeMethod(
          testManager,
          testClass,
          'aMethod',
          argThat(isEmpty),
        ),
        testManager.converter.mock.convertForRemoteManager(
          testManager,
          argThat(isNull),
        ),
      ]);
    });

    test('invokeLocalMethodOnUnpairedReference', () {
      when(testManager.localHandler.createInstance(testManager, TestClass, any))
          .thenReturn(TestClass());

      testManager.onReceiveInvokeMethodOnUnpairedReference(
        UnpairedReference(0, <Object>[]),
        'aMethod',
      );

      verify(
        testManager.localHandler.sendInvokeMethod(
          testManager,
          argThat(isA<TestClass>()),
          'aMethod',
          argThat(isEmpty),
        ),
      );
    });

    test('disposePairWithRemoteReference', () {
      when(testManager.localHandler.createInstance(testManager, TestClass, any))
          .thenReturn(TestClass());

      final TestClass testClass = testManager.onReceiveCreateNewPair(
        RemoteReference('tea'),
        0,
      );

      testManager.onReceiveDisposePair(RemoteReference('tea'));

      verify(
          testManager.localHandler.onInstanceDisposed(testManager, testClass));
      expect(
        testManager.getPairedLocalReference(RemoteReference('tea')),
        isNull,
      );
      expect(testManager.getPairedRemoteReference(testClass), isNull);
    });

    test('pairWithNewRemoteReference', () async {
      final TestClass testClass = TestClass();

      when(testManager.remoteHandler.getCreationArguments(any))
          .thenReturn(<Object>[]);

      final RemoteReference remoteReference =
          await testManager.pairWithNewRemoteReference(
        testClass,
      );

      expect(remoteReference, isNotNull);
      expect(testManager.getPairedLocalReference(remoteReference), testClass);
      expect(testManager.getPairedRemoteReference(testClass), remoteReference);
      verifyInOrder([
        testManager.converter.mock.convertForRemoteManager(
          testManager,
          argThat(isEmpty),
        ),
        testManager.remoteHandler
            .createInstance(remoteReference, 0, argThat(isEmpty)),
      ]);
    });

    test('pairWithNewRemoteReference returns null', () async {
      final TestClass testClass = TestClass();

      when(testManager.remoteHandler.getCreationArguments(any))
          .thenReturn(<Object>[]);

      testManager.pairWithNewRemoteReference(testClass);
      expect(
        testManager.pairWithNewRemoteReference(testClass),
        completion(isNull),
      );
    });

    test('invokeRemoteStaticMethod', () async {
      await testManager.invokeStaticMethod(TestClass, 'aStaticMethod');
      verifyInOrder([
        testManager.converter.mock.convertForRemoteManager(
          testManager,
          argThat(isNull),
        ),
        testManager.remoteHandler.sendInvokeStaticMethod(
          0,
          'aStaticMethod',
          argThat(isEmpty),
        ),
        testManager.converter.mock.convertForLocalManager(
          testManager,
          argThat(isNull),
        ),
      ]);
    });

    test('invokeRemoteMethod', () async {
      final testClass = TestClass();
      testManager.pairWithNewRemoteReference(testClass);

      when(testManager.remoteHandler.getCreationArguments(any))
          .thenReturn(<Object>[]);

      await testManager.invokeMethod(
        testManager.getPairedRemoteReference(testClass),
        'aMethod',
      );

      verifyInOrder([
        testManager.converter.mock.convertForRemoteManager(
          testManager,
          argThat(isNull),
        ),
        testManager.remoteHandler.sendInvokeMethod(
          testManager.getPairedRemoteReference(testClass),
          'aMethod',
          argThat(isEmpty),
        ),
      ]);
    });

    test(
      'invokeRemoteMethodOnUnpairedReference',
      () async {
        when(testManager.remoteHandler.getCreationArguments(any))
            .thenReturn(<Object>[]);

        await testManager.invokeMethodOnUnpairedReference(
          TestClass(),
          'aMethod',
        );

        verify(testManager.converter.convertForRemoteManager(
          testManager,
          any,
        )).called(2);
        verify(testManager.remoteHandler.sendInvokeMethodOnUnpairedReference(
          argThat(isUnpairedReference(0, <Object>[], null)),
          'aMethod',
          argThat(isEmpty),
        ));
      },
    );

    test('disposePairWithLocalReference', () async {
      final testClass = TestClass();

      final RemoteReference remoteReference =
          await testManager.pairWithNewRemoteReference(
        testClass,
      );
      testManager.disposePair(testClass);

      expect(testManager.getPairedLocalReference(remoteReference), isNull);
      expect(testManager.getPairedRemoteReference(testClass), isNull);
    });
  });

  group('$PoolableReferencePairManager', () {
    ReferencePairManagerPool pool;
    TestPoolableReferencePairManager testManager1;
    TestPoolableReferencePairManager testManager2;

    setUp(() {
      pool = ReferencePairManagerPool();
      testManager1 = TestPoolableReferencePairManager(<Type>[TestClass], 'id1')
        ..initialize();
      testManager2 = TestPoolableReferencePairManager(<Type>[TestClass2], 'id2')
        ..initialize();
    });

    test('add', () {
      expect(pool.add(testManager1), isTrue);
      expect(pool.add(testManager1), isTrue);
      expect(
        pool.add(TestPoolableReferencePairManager(<Type>[TestClass], 'id3')),
        isFalse,
      );
      expect(
        pool.add(TestPoolableReferencePairManager(<Type>[TestClass2], 'id1')),
        isFalse,
      );
      expect(pool.add(testManager2), isTrue);
    });

    test('remove', () {
      pool.add(testManager1);
      pool.remove(testManager1);

      expect(
        pool.add(TestPoolableReferencePairManager(<Type>[TestClass], 'id1')),
        isTrue,
      );
    });
  });
}

class TestClass with LocalReference {
  @override
  Type get referenceType => TestClass;
}

class TestClass2 extends TestClass {
  @override
  Type get referenceType => TestClass2;
}

class TestReferencePairManager extends RemoteReferenceMap {
  TestReferencePairManager() : super(<Type>[TestClass]);

  @override
  final MockLocalHandler localHandler = MockLocalHandler();

  @override
  final MockRemoteHandler remoteHandler = MockRemoteHandler();

  @override
  final SpyReferenceConverter converter = SpyReferenceConverter();
}

class TestPoolableReferencePairManager extends PoolableReferencePairManager {
  TestPoolableReferencePairManager(
    List<Type> supportedTypes,
    String poolId,
  ) : super(supportedTypes, poolId);

  @override
  final MockLocalHandler localHandler = MockLocalHandler();

  @override
  final MockRemoteHandler remoteHandler = MockRemoteHandler();
}

class MockRemoteHandler extends Mock implements MessageSender {}

class MockLocalHandler extends Mock implements ReferenceChannelHandler {}

class MockReferenceConverter extends Mock implements ReferenceConverter {}

class SpyReferenceConverter extends StandardReferenceConverter {
  SpyReferenceConverter();

  final ReferenceConverter mock = MockReferenceConverter();

  @override
  Object convertForRemoteManager(
    RemoteReferenceMap manager,
    Object object,
  ) {
    mock.convertForRemoteManager(manager, object);
    return super.convertForRemoteManager(manager, object);
  }

  @override
  Object convertForLocalManager(
    RemoteReferenceMap manager,
    Object object,
  ) {
    mock.convertForLocalManager(manager, object);
    return super.convertForLocalManager(manager, object);
  }
}

File: reference/android/src/test/java/github/penguin/reference/ReferenceTest.java
package github.penguin.reference;

import org.junit.Before;
import org.junit.Test;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;
import java.util.Collections;
import java.util.List;
import github.penguin.reference.async.Completable;
import github.penguin.reference.async.Completer;
import github.penguin.reference.reference.LocalReference;
import github.penguin.reference.reference.PoolableReferencePairManager;
import github.penguin.reference.reference.ReferenceConverter;
import github.penguin.reference.reference.ReferenceConverter.StandardReferenceConverter;
import github.penguin.reference.reference.ReferencePairManager;
import github.penguin.reference.reference.ReferencePairManagerPool;
import github.penguin.reference.reference.RemoteReference;
import github.penguin.reference.reference.UnpairedReference;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.ArgumentMatchers.isNull;
import static org.mockito.ArgumentMatchers.notNull;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.mockito.Mockito.any;

@SuppressWarnings("rawtypes")
public class ReferenceTest {
  private static TestReferencePairManager testManager;
  private static TestPoolableReferencePairManager testPoolableManager1;
  private static TestPoolableReferencePairManager testPoolableManager2;
  private static ReferencePairManagerPool pool;

  private static class TestClass implements LocalReference {
    @Override
    public Class<? extends LocalReference> getReferenceClass() {
      return TestClass.class;
    }
  }

  private static class TestClass2 extends TestClass {
    @Override
    public Class<? extends LocalReference> getReferenceClass() {
      return TestClass2.class;
    }
  }

  private static class TestReferencePairManager extends ReferencePairManager {
    private final LocalReferenceCommunicationHandler localHandler;
    private final RemoteReferenceCommunicationHandler remoteHandler;
    private final StandardReferenceConverter converter;

    private TestReferencePairManager() {
      super(Collections.<Class<? extends LocalReference>>singletonList(TestClass.class));
      this.localHandler = mock(LocalReferenceCommunicationHandler.class);
      this.remoteHandler = mock(RemoteReferenceCommunicationHandler.class);
      this.converter = spy(new StandardReferenceConverter());
    }

    @Override
    public RemoteReferenceCommunicationHandler getRemoteHandler() {
      return remoteHandler;
    }

    @Override
    public LocalReferenceCommunicationHandler getLocalHandler() {
      return localHandler;
    }

    @Override
    public ReferenceConverter getConverter() {
      return converter;
    }
  }

  private static class TestPoolableReferencePairManager extends PoolableReferencePairManager {
    private final LocalReferenceCommunicationHandler localHandler;
    private final RemoteReferenceCommunicationHandler remoteHandler;

    private TestPoolableReferencePairManager(List<Class<? extends LocalReference>> supportedClasses, String poolId) {
      super(supportedClasses, poolId);
      this.localHandler = mock(LocalReferenceCommunicationHandler.class);
      this.remoteHandler = mock(RemoteReferenceCommunicationHandler.class);
    }

    @Override
    public RemoteReferenceCommunicationHandler getRemoteHandler() {
      return remoteHandler;
    }

    @Override
    public LocalReferenceCommunicationHandler getLocalHandler() {
      return localHandler;
    }
  }

  @Before
  public void setUp() {
    testManager = new TestReferencePairManager();
    testManager.initialize();

    pool = new ReferencePairManagerPool();

    testPoolableManager1 = new TestPoolableReferencePairManager(
        Collections.<Class<? extends LocalReference>>singletonList(TestClass.class),
        "id1");
    testPoolableManager1.initialize();

    testPoolableManager2 = new TestPoolableReferencePairManager(
        Collections.<Class<? extends LocalReference>>singletonList(TestClass2.class),
        "id2");
    testPoolableManager2.initialize();
  }

  @Test
  public void referencePairManager_pairWithNewLocalReference() throws Exception {
    when(testManager.localHandler.create(eq(testManager), eq(TestClass.class), anyList()))
        .thenReturn(new TestClass());

    final TestClass result = (TestClass) testManager.pairWithNewLocalReference(
        new RemoteReference("apple"),
        0
    );

    assertEquals(testManager.getPairedLocalReference(new RemoteReference("apple")), result);
    assertEquals(testManager.getPairedRemoteReference(result), new RemoteReference("apple"));

    verify(testManager.converter).convertReferencesForLocalManager(eq(testManager), anyList());
    verify(testManager.localHandler).create(eq(testManager), eq(TestClass.class), anyList());
  }

  @Test
  public void referencePairManager_pairWithNewLocalReference_returnsNull() throws Exception {
    when(testManager.localHandler.create(eq(testManager), eq(TestClass.class), anyList()))
        .thenReturn(new TestClass());

    testManager.pairWithNewLocalReference(new RemoteReference("apple"), 0);
    assertNull(testManager.pairWithNewLocalReference(new RemoteReference("apple"), 0));
  }

  @Test
  public void referencePairManager_invokeLocalMethod() throws Exception {
    when(testManager.localHandler.create(eq(testManager), eq(TestClass.class), anyList()))
        .thenReturn(new TestClass());

    final LocalReference localReference =
        testManager.pairWithNewLocalReference(new RemoteReference("chi") , 0);
    reset(testManager.converter);

    testManager.invokeLocalMethod(localReference, "aMethod");

    verify(testManager.converter).convertReferencesForLocalManager(eq(testManager), anyList());
    verify(testManager.localHandler).invokeMethod(eq(testManager),
        eq(localReference),
        eq("aMethod"),
        anyList()
    );
    verify(testManager.converter).convertReferencesForRemoteManager(eq(testManager), isNull());
  }

  @Test
  public void referencePairManager_invokeLocalMethodOnUnpairedReference() throws Exception {
    when(testManager.localHandler.create(eq(testManager), eq(TestClass.class), anyList()))
        .thenReturn(new TestClass());

    testManager.invokeLocalMethodOnUnpairedReference(new UnpairedReference(0, Collections.emptyList()), "aMethod");

    verify(testManager.localHandler).invokeMethod(eq(testManager), (LocalReference) notNull(), eq("aMethod"), anyList());
  }

  @Test
  public void referencePairManager_disposePairWithRemoteReference() throws Exception {
    when(testManager.localHandler.create(eq(testManager), eq(TestClass.class), anyList()))
        .thenAnswer(new Answer<Object>() {
          @Override
          public Object answer(InvocationOnMock invocation) {
            return new TestClass();
          }
        });

    final TestClass testClass = (TestClass) testManager.pairWithNewLocalReference(new RemoteReference("tea") ,0);
    testManager.disposePairWithRemoteReference(new RemoteReference("tea"));

    verify(testManager.localHandler).dispose(testManager, testClass);
    assertNull(testManager.getPairedLocalReference(new RemoteReference("tea")));
    assertNull(testManager.getPairedRemoteReference(testClass));
  }

  @SuppressWarnings("unchecked")
  @Test
  public void referencePairManager_pairWithNewRemoteReference() {
    final TestClass testClass = new TestClass();

    when(testManager.remoteHandler.getCreationArguments(testClass))
        .thenReturn(Collections.emptyList());

    when(testManager.remoteHandler.create(any(RemoteReference.class), eq(0), anyList()))
        .thenReturn(new Completer<Void>().complete(null).completable);

    final Completable.OnCompleteListener mockListener = mock(Completable.OnCompleteListener.class);
    testManager.pairWithNewRemoteReference(testClass).setOnCompleteListener(mockListener);
    final RemoteReference remoteReference = testManager.getPairedRemoteReference(testClass);

    verify(mockListener).onComplete(remoteReference);
    assertEquals(testManager.getPairedLocalReference(remoteReference), testClass);
    assertEquals(testManager.getPairedRemoteReference(testClass), remoteReference);

    verify(testManager.converter).convertReferencesForRemoteManager(eq(testManager), anyList());
    verify(testManager.remoteHandler).create(eq(remoteReference), eq(0), anyList());
  }

  @SuppressWarnings("unchecked")
  @Test
  public void referencePairManager_pairWithNewRemoteReference_returnsNull() {
    final TestClass testClass = new TestClass();

    when(testManager.remoteHandler.getCreationArguments(testClass))
        .thenReturn(Collections.emptyList());

    when(testManager.remoteHandler.create(any(RemoteReference.class), eq(0), anyList()))
        .thenReturn(new Completer<Void>().complete(null).completable);

    testManager.pairWithNewRemoteReference(testClass);

    final Completable.OnCompleteListener mockListener = mock(Completable.OnCompleteListener.class);
    testManager.pairWithNewRemoteReference(testClass).setOnCompleteListener(mockListener);
    verify(mockListener).onComplete(null);
  }

  @Test
  public void referencePairManager_invokeRemoteMethod() throws Exception {
    final TestClass testClass = new TestClass();

    when(testManager.remoteHandler.create(any(RemoteReference.class), eq(0), anyList()))
        .thenReturn(new Completer<Void>().complete(null).completable);

    when(testManager.remoteHandler.invokeMethod(any(RemoteReference.class), eq("aMethod"), anyList()))
        .thenReturn(new Completer<>().complete(null).completable);

    testManager.pairWithNewRemoteReference(testClass);
    reset(testManager.converter);

    final RemoteReference remoteReference = testManager.getPairedRemoteReference(testClass);
    testManager.invokeRemoteMethod(remoteReference, "aMethod");

    verify(testManager.converter).convertReferencesForRemoteManager(eq(testManager), anyList());
    verify(testManager.remoteHandler).invokeMethod(eq(remoteReference), eq("aMethod"), anyList());
    verify(testManager.converter).convertReferencesForLocalManager(eq(testManager), isNull());
  }

  @Test
  public void referencePairManager_invokeRemoteMethodOnUnpairedReference() throws Exception {
    when(testManager.remoteHandler.create(any(RemoteReference.class), eq(0), anyList()))
        .thenReturn(new Completer<Void>().complete(null).completable);

    when(testManager.remoteHandler.invokeMethodOnUnpairedReference(any(UnpairedReference.class), eq("aMethod"), anyList()))
        .thenReturn(new Completer<>().complete(null).completable);

    testManager.invokeRemoteMethodOnUnpairedReference(new TestClass(), "aMethod");

    verify(testManager.converter, times(2)).convertReferencesForRemoteManager(eq(testManager), anyList());
    verify(testManager.remoteHandler).invokeMethodOnUnpairedReference(any(UnpairedReference.class),
        eq("aMethod"),
        anyList()
    );
    verify(testManager.converter).convertReferencesForLocalManager(eq(testManager), isNull());
  }

  @Test
  public void referencePairManager_disposePairWithLocalReference() {
    final TestClass testClass = new TestClass();

    when(testManager.remoteHandler.create(any(RemoteReference.class), eq(0), anyList()))
        .thenReturn(new Completer<Void>().complete(null).completable);

    testManager.pairWithNewRemoteReference(testClass);
    final RemoteReference remoteReference = testManager.getPairedRemoteReference(testClass);

    testManager.disposePairWithLocalReference(testClass);

    verify(testManager.remoteHandler).dispose(remoteReference);

    assertNull(testManager.getPairedLocalReference(remoteReference));
    assertNull(testManager.getPairedRemoteReference(testClass));
  }

  @Test
  public void poolableReferencePairManager_add() {
    final PoolableReferencePairManager sameClassManager = new TestPoolableReferencePairManager(
        Collections.<Class<? extends LocalReference>>singletonList(TestClass.class),
        "id3");

    final PoolableReferencePairManager sameIdManager = new TestPoolableReferencePairManager(
        Collections.<Class<? extends LocalReference>>singletonList(TestClass2.class),
        "id1");

    assertTrue(pool.add(testPoolableManager1));
    assertTrue(pool.add(testPoolableManager1));
    assertFalse(pool.add(sameClassManager));
    assertFalse(pool.add(sameIdManager));
    assertTrue(pool.add(testPoolableManager2));
  }

  @Test
  public void poolableReferencePairManager_remove() {
    pool.add(testPoolableManager1);
    pool.remove(testPoolableManager1);

    final PoolableReferencePairManager manager = new TestPoolableReferencePairManager(
        Collections.<Class<? extends LocalReference>>singletonList(TestClass.class),
        "id1");

    assertTrue(pool.add(manager));
  }
}
